dodo.py

```diff
@@ -0,0 +1,325 @@
+    """Generate JSON files that will be used for syntax highlighting."""
+    script = 'highlighting/generate-rules-highlighting.py'
+
+    return dict(
+        title=lambda _: 'Generate highlighting rules',
+        file_dep=[
+            *glob('highlighting/*.py'),
+            script,
+        ],
+        actions=[
+            [python3, script],
+        ],
+        targets=['highlighting/highlighting.json', 'highlighting/highlighting-trad.json'],
+    )
+
+
+def task_compile_babel():
+    """Compile .po files for use with Babel."""
+    return dict(
+        title=lambda _: 'Compile Babel files',
+        file_dep=pofiles,
+        actions=[
+            'pybabel compile -f -d translations',
+        ],
+        # Every .po file leads to a .mo file
+        targets=mofiles,
+
+        verbosity=0,  # stderr is too noisy by default
+    )
+
+
+def task_generate_static_babel_content():
+    """Extract information from Babel at build time.
+
+    We do this so we don't have to keep Babel's locale database in memory at
+    runtime.
+    """
+    script = 'build-tools/heroku/generate-static-babel-content'
+    return dict(
+        title=lambda _: 'Generate static Babel content',
+        file_dep=[script],
+        actions=[
+            [python3, script],
+        ],
+        targets=['static_babel_content.json'],
+        uptodate=[babel_version_unchanged],
+    )
+
+
+def task_client_messages():
+    """Use Python to extract messages to a .ts file that will be used in the frontend."""
+    script = 'build-tools/heroku/generate-client-messages.py'
+
+    return dict(
+        title=lambda _: 'Generate client messages',
+
+        # Depends on a specific YAML file and the .mo files
+        file_dep=[
+            *glob('content/client-messages/*.yaml'),
+            *mofiles,
+            script,
+        ],
+        actions=[
+            [python3, script],
+        ],
+        targets=[
+            'static/js/message-translations.ts',
+        ],
+    )
+
+
+def task_typescript():
+    """Compile typescript."""
+    return dict(
+        file_dep=[
+            *glob('static/js/**/*.ts', recursive=True),
+            # Some non-.ts files that we also own
+            'static/js/buttons.js',
+            'static/js/skulpt_debugger.js',
+        ],
+        task_dep=['generate_highlighting', 'client_messages'],
+        title=lambda _: 'Compile TypeScript',
+        actions=[
+            # Use tsc to generate .js (this downlevels to old JavaScript versions for old browsers)
+            ['npx', 'tsc', '--outDir', '__tmp__'],
+
+            # Then bundle JavaScript into a single bundle
+            ['npx', 'esbuild', '__tmp__/static/js/index.js',
+             '--bundle', '--sourcemap', '--minify', '--target=es2017',
+             '--global-name=hedyApp', '--platform=browser',
+             '--outfile=static/js/appbundle.js'],
+
+            # Delete tempdir
+            ['rm', '-rf', '__tmp__'],
+        ],
```

Comment:
I personally don't like this syntax. I think having this command as a string and splitting it into multiple lines (if needed) would be more readable. And in the case of including variables, we could simply use string interpolation!

Reply:
> And in the case of including variables, we could simply use string interpolation!

Yeah but here's the thing: you need to escape special characters inside those strings, otherwise they get pass uninterpreted to the shell and the command might do something else than you want.

A trivial example that will probably only fail:

```py

def copy_file(file1, file2):
  run_command(f'cp {file1} {file2}')  # Seems easy enough

# This is fine
copy_file('one.txt', 'two.txt')  # cp one.txt two.txt

# This is not
copy_file('file with spaces.txt', 'two.txt')   # cp file with spaces.txt two.txt
# ^^^ oops! Now the cp command gets 4 arguments instead of 2!
```

Not to mention special characters like `&` and `>` and `;` etc, that can make the shell do weird things.

```py
copy_file('file1.txt', 'file2.txt; rm -rf /')    # cp file1.txt file2.txt; rm -rf / 
                                        # ^^^^ the shell will happily execute this
```

I guess for the domain of a build tool, where we don't really handle untrusted strings, it matters less. But I've learned to be hyper-allergic to strings for shell commands, for this reason.